processing/src/main/java/io/druid/query/filter/BoundDimFilter.java

```diff
@@ -343,83 +341,197 @@ public boolean applyLong(long input)
               return false;
             }
           };
+        } else {
+          return predicate;
+        }
+      }
+
+      private void initPredicate()
```

Comment:
Actually I'm losing the point of lazy initialization here, as well several similar places below. Maybe it's simpler to compute build a predicate eagerly.